devdoc/job_queue.md

```diff
@@ -0,0 +1,59 @@
+# `job_queue` requirements
```

Comment:
I see now you mentioned on teams that this is for parked calls. Is there a larger design that this fits into? I think seeing some example or the big picture of how this module supports parked calls might help, perhaps in a separate design doc.

Reply:
Also, when I did a parked calls implementation, I ran into the following problem and I'm curious if your overall design solves this.

In my case, I have a module that resolves SF client endpoints or returns the client if it's already resolved. When a user calls "get_client", if there is no client yet, it would push the request onto a mpsc queue, parking the call until the client is resolved by another worker thread.

Then a single worker thread would resolve the SF client, and pop all the items from the queue, calling their callbacks with the client. If that client resolution failed, it still popped the items and called the callbacks with error.

The callers had retry logic on error to call "get_client" again, which would push back into the queue. This caused infinite loops because the queue would never fully drain.

The solution I used was to insert a barrier into the queue in the worker thread before draining the queue and to not handle anything queued after the barrier (because the newer requests would again trigger the worker thread to retry when its current iteration was complete).

Reply:
I haven't run into the infinite loop problem yet. I'll keep that in mind.